src/main/java/com/es/phoneshop/model/service/CartService.java

```diff
@@ -24,26 +23,47 @@ public static CartService getInstance() {
     }
 
     public synchronized void add(HttpSession session, Product product, int quantity) throws OutOfStockException {
-        if (session.getAttribute(SESSION_ATTRIBUTE_CART) == null) {
-            session.setAttribute(SESSION_ATTRIBUTE_CART, new Cart());
-        }
-        Cart cart = (Cart) session.getAttribute(SESSION_ATTRIBUTE_CART);
+        Cart cart = extractCartOrCreateNewOne(session);
         int cartAmount = cart.getCartItemList().stream()
                 .filter(p -> p.getProduct().getId().equals(product.getId()))
                 .findAny()
                 .map(CartItem::getQuantity)
                 .orElse(0);
-        if (product.getStock() - cartAmount < quantity || quantity < 0) {
+        if (product.getStock() - cartAmount < quantity || quantity <= 0) {
```

Comment:
You might actually remove product.getStock() - cartAmount < quantity check from here as it is already performed in update method (so, leave only negative quantity check in this method)